# Importing libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn

# Importing and cleaning of dataset

In [2]:
data = pd.read_csv("Training.csv")
data.head()

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          1                     1                    0          0   
2        1          0                     1                    0          0   
3        1          1                     0                    0          0   
4        1          1                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  scurring  \
0       0           0             0        0                 0  ...         0   
1       0           0             0        0                 0  ...         0   
2       0           0             0        0                 0  ...         0   
3       0           0             0        0                 0  ...         0   
4       0           0             0        0                 0  ...         0   

   skin_peeling  silver_like_dusting  small_dents_in_nails  \
0             0                    0                     0   
1             0                    0                     0   
2             0                    0                     0   
3             0                    0                     0   
4             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   
3                   0        0                     0                  0   
4                   0        0                     0                  0   

          prognosis  Unnamed: 133  
0  Fungal infection           NaN  
1  Fungal infection           NaN  
2  Fungal infection           NaN  
3  Fungal infection           NaN  
4  Fungal infection           NaN  

[5 rows x 134 columns]

In [3]:
## removing the last column of the dataframe
data.drop(data.columns[-1], axis=1,inplace=True)

In [4]:
data.head(1)

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0       0           0             0        0                 0  ...   

   blackheads  scurring  skin_peeling  silver_like_dusting  \
0           0         0             0                    0   

   small_dents_in_nails  inflammatory_nails  blister  red_sore_around_nose  \
0                     0                   0        0                     0   

   yellow_crust_ooze         prognosis  
0                  0  Fungal infection  

[1 rows x 133 columns]

In [5]:
data.drop_duplicates(inplace=True)
data = data.reset_index()

In [6]:
data.shape

(304, 134)

In [7]:
data.head(3)

index  itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  \
0      0        1          1                     1                    0   
1      1        0          1                     1                    0   
2      2        1          0                     1                    0   

   shivering  chills  joint_pain  stomach_pain  acidity  ...  blackheads  \
0          0       0           0             0        0  ...           0   
1          0       0           0             0        0  ...           0   
2          0       0           0             0        0  ...           0   

   scurring  skin_peeling  silver_like_dusting  small_dents_in_nails  \
0         0             0                    0                     0   
1         0             0                    0                     0   
2         0             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   

          prognosis  
0  Fungal infection  
1  Fungal infection  
2  Fungal infection  

[3 rows x 134 columns]

# Data exploration

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 304 entries, 0 to 303
Columns: 134 entries, index to prognosis
dtypes: int64(133), object(1)
memory usage: 318.4+ KB


In [9]:
data.describe()

index     itching   skin_rash  nodal_skin_eruptions  \
count  304.000000  304.000000  304.000000            304.000000   
mean   205.855263    0.134868    0.138158              0.013158   
std    111.641719    0.342146    0.345635              0.114139   
min      0.000000    0.000000    0.000000              0.000000   
25%    115.750000    0.000000    0.000000              0.000000   
50%    206.500000    0.000000    0.000000              0.000000   
75%    300.500000    0.000000    0.000000              0.000000   
max    407.000000    1.000000    1.000000              1.000000   

       continuous_sneezing   shivering      chills  joint_pain  stomach_pain  \
count           304.000000  304.000000  304.000000  304.000000    304.000000   
mean              0.039474    0.013158    0.174342    0.151316      0.036184   
std               0.195040    0.114139    0.380029    0.358947      0.187056   
min               0.000000    0.000000    0.000000    0.000000      0.000000   
25%               0.000000    0.000000    0.000000    0.000000      0.000000   
50%               0.000000    0.000000    0.000000    0.000000      0.000000   
75%               0.000000    0.000000    0.000000    0.000000      0.000000   
max               1.000000    1.000000    1.000000    1.000000      1.000000   

          acidity  ...  pus_filled_pimples  blackheads    scurring  \
count  304.000000  ...          304.000000  304.000000  304.000000   
mean     0.049342  ...            0.013158    0.013158    0.013158   
std      0.216938  ...            0.114139    0.114139    0.114139   
min      0.000000  ...            0.000000    0.000000    0.000000   
25%      0.000000  ...            0.000000    0.000000    0.000000   
50%      0.000000  ...            0.000000    0.000000    0.000000   
75%      0.000000  ...            0.000000    0.000000    0.000000   
max      1.000000  ...            1.000000    1.000000    1.000000   

       skin_peeling  silver_like_dusting  small_dents_in_nails  \
count    304.000000           304.000000            304.000000   
mean       0.019737             0.019737              0.019737   
std        0.139324             0.139324              0.139324   
min        0.000000             0.000000              0.000000   
25%        0.000000             0.000000              0.000000   
50%        0.000000             0.000000              0.000000   
75%        0.000000             0.000000              0.000000   
max        1.000000             1.000000              1.000000   

       inflammatory_nails     blister  red_sore_around_nose  yellow_crust_ooze  
count          304.000000  304.000000            304.000000         304.000000  
mean             0.019737    0.016447              0.016447           0.016447  
std              0.139324    0.127398              0.127398           0.127398  
min              0.000000    0.000000              0.000000           0.000000  
25%              0.000000    0.000000              0.000000           0.000000  
50%              0.000000    0.000000              0.000000           0.000000  
75%              0.000000    0.000000              0.000000           0.000000  
max              1.000000    1.000000              1.000000           1.000000  

[8 rows x 133 columns]

In [10]:
data.isnull().any().to_numpy()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

## Data exploration results : 
- Shape : (304,133)
- Dtypes : int64(132), object(1)
- Memory usage : 318.2 KB
- No duplicate rows
- No Nan values

# Data preprocessing

##### Converting categorical data into numerical data
###### 1. Method 1 : Dummy variable (OneHotEncoding)
###### 2. Method 2 : Label Encoding
- In this project we're going to use label encoding

In [11]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [12]:
le_data = pd.DataFrame(le.fit_transform(data[["prognosis"]]), columns=["Disease"])
le_data.tail(5)

E:\Anaconda\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Disease
299       27
300       27
301       27
302       27
303       27

In [13]:
len(le_data)

304

In [14]:
len(data)

304

In [62]:
merged_data = pd.concat([data,le_data],axis=1)

In [16]:
merged_data.tail()

index  itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  \
299    402        0          0                     0                    0   
300    403        0          1                     0                    0   
301    405        0          1                     0                    0   
302    406        0          1                     0                    0   
303    407        0          1                     0                    0   

     shivering  chills  joint_pain  stomach_pain  acidity  ...  scurring  \
299          0       0           0             0        0  ...         0   
300          0       0           0             0        0  ...         0   
301          0       0           0             0        0  ...         0   
302          0       0           0             0        0  ...         0   
303          0       0           0             0        0  ...         0   

     skin_peeling  silver_like_dusting  small_dents_in_nails  \
299             0                    0                     0   
300             0                    0                     0   
301             0                    0                     0   
302             0                    0                     0   
303             0                    0                     0   

     inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
299                   0        1                     1                  1   
300                   0        1                     1                  1   
301                   0        0                     1                  1   
302                   0        1                     0                  1   
303                   0        1                     1                  0   

     prognosis  Disease  
299   Impetigo       27  
300   Impetigo       27  
301   Impetigo       27  
302   Impetigo       27  
303   Impetigo       27  

[5 rows x 135 columns]

In [63]:
merged_data = merged_data.drop("index", axis=1)

# Creating csv file for disease and prognosis

In [24]:
disease_prognosis = pd.DataFrame(merged_data[["prognosis", "Disease"]])
disease_prognosis.head(3)

prognosis  Disease
0  Fungal infection       15
1  Fungal infection       15
2  Fungal infection       15

In [25]:
disease_prognosis.drop_duplicates(inplace=True)
disease_prognosis = disease_prognosis.reset_index()
len(disease_prognosis)

41

In [26]:
disease_prognosis.drop("index", axis=1, inplace=True)

In [27]:
disease_prognosis.head()

prognosis  Disease
0     Fungal infection       15
1              Allergy        4
2                 GERD       16
3  Chronic cholestasis        9
4        Drug Reaction       14

In [28]:
disease_prognosis.to_csv("Disease Prognosis.csv")

# Algorithm 1 : Logistic Regression

In [68]:
X_train = merged_data.drop(["prognosis", "Disease"], axis=1)
y_train = merged_data.Disease

In [69]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X_train,y_train)

LogisticRegression()

### Loading test data

In [49]:
test_data = pd.read_csv("Testing.csv")
len(test_data)

42

In [52]:
test_data.head()

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          0                     0                    1          1   
2        0          0                     0                    0          0   
3        1          0                     0                    0          0   
4        1          1                     0                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  \
0       0           0             0        0                 0  ...   
1       1           0             0        0                 0  ...   
2       0           0             1        1                 1  ...   
3       0           0             0        0                 0  ...   
4       0           0             1        0                 0  ...   

   blackheads  scurring  skin_peeling  silver_like_dusting  \
0           0         0             0                    0   
1           0         0             0                    0   
2           0         0             0                    0   
3           0         0             0                    0   
4           0         0             0                    0   

   small_dents_in_nails  inflammatory_nails  blister  red_sore_around_nose  \
0                     0                   0        0                     0   
1                     0                   0        0                     0   
2                     0                   0        0                     0   
3                     0                   0        0                     0   
4                     0                   0        0                     0   

   yellow_crust_ooze            prognosis  
0                  0     Fungal infection  
1                  0              Allergy  
2                  0                 GERD  
3                  0  Chronic cholestasis  
4                  0        Drug Reaction  

[5 rows x 133 columns]

### Label encoding the test data

In [57]:
test_le = pd.DataFrame(le.fit_transform(test_data.prognosis), columns=["Disease"])
test_data = pd.concat([test_data,test_le], axis=1)

In [59]:
test_data.head(3)

itching  skin_rash  nodal_skin_eruptions  continuous_sneezing  shivering  \
0        1          1                     1                    0          0   
1        0          0                     0                    1          1   
2        0          0                     0                    0          0   

   chills  joint_pain  stomach_pain  acidity  ulcers_on_tongue  ...  scurring  \
0       0           0             0        0                 0  ...         0   
1       1           0             0        0                 0  ...         0   
2       0           0             1        1                 1  ...         0   

   skin_peeling  silver_like_dusting  small_dents_in_nails  \
0             0                    0                     0   
1             0                    0                     0   
2             0                    0                     0   

   inflammatory_nails  blister  red_sore_around_nose  yellow_crust_ooze  \
0                   0        0                     0                  0   
1                   0        0                     0                  0   
2                   0        0                     0                  0   

          prognosis  Disease  
0  Fungal infection       15  
1           Allergy        4  
2              GERD       16  

[3 rows x 134 columns]

### Score of the model

In [60]:
X_test = test_data.drop(["prognosis", "Disease"], axis=1)
y_test = test_data.Disease

In [66]:
lr.score(X_test, y_test)

1.0

# Algorithm 2 : SVM

In [71]:
from sklearn.svm import SVC
svm = SVC()
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

1.0

# Algorithm 3 : RandomForest

In [72]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf.fit(X_train,y_train)
rf.score(X_test, y_test)

0.9761904761904762

# Algorithm 4 : Decision Tree

In [73]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dt.fit(X_train,y_train)
dt.score(X_test, y_test)

0.9761904761904762

# Algorithm 5 : Naive Bayes

In [74]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(X_train,y_train)
gnb.score(X_test, y_test)

1.0

*********************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

### Score for different algorithms are :
- 1. LogisticRegression : 1.0
- 2. SVM : 1.0
- 3. RandomForest : 0.9761904761904762
- 4. DecisionTree : 0.9761904761904762
- 5. NaiveBayes : 1.0

***************************************************************************************************************************